File: Zds/Zds.Flow/Collections/SafeList.vb
Namespace Collections
    <DebuggerStepThrough>
    Public Class SafeList(Of T)
        Private List As New List(Of T)
        Public ReadOnly Property Elements As T() = New T(-1) {}
        Public ReadOnly Property Count As Integer
            Get
                Return Elements.Length
            End Get
        End Property
        Public Sub Add(Item As T)
            SyncLock List
                List.Add(Item)
                UpdateCache()
            End SyncLock
        End Sub
        Public Sub Remove(Item As T)
            SyncLock List
                If List.Remove(Item) Then UpdateCache()
            End SyncLock
        End Sub
        Public Sub Add(Items As T())
            SyncLock List
                List.AddRange(Items)
                UpdateCache()
            End SyncLock
        End Sub
        Public Sub Remove(Items As T())
            SyncLock List
                For Each Element In Items
                    List.Remove(Element)
                Next
                UpdateCache()
            End SyncLock
        End Sub
        Public Sub Clear()
            SyncLock List
                List.Clear()
                UpdateCache()
            End SyncLock
        End Sub
        Public Function Contains(Obj As T) As Boolean
            Return Elements.Contains(Obj)
        End Function
        Public Sub UpdateCache()
            SyncLock List
                _Elements = List.ToArray
            End SyncLock
        End Sub
    End Class
End Namespace

File: Flow/Flow/Collections/SafeList.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;

namespace Flow.Collections
{
    public class SafeList<T> : IList<T>
    {
        private readonly List<T> list;
        private T[] cache;
        public T[] Elements => cache;
        public T this[int index]
        {
            get => cache[index];
            set
            {
                lock (list)
                {
                    list[index] = value;
                    UpdateCache();
                }
            }
        }
        public object Lock => list;
        public int Count => cache.Length;
        public bool IsReadOnly => false;
        public void Add(T item)
        {
            lock (list)
            {
                list.Add(item);
                UpdateCache();
            }
        }
        public void AddRange(T[] items)
        {
            lock (list)
            {
                list.AddRange(items);
                UpdateCache();
            }
        }
        public void Insert(int index, T item)
        {
            lock (list)
            {
                list.Insert(index, item);
                UpdateCache();
            }
        }
        public bool Remove(T item)
        {
            lock (list)
            {
                bool output = list.Remove(item);
                UpdateCache();
                return output;
            }
        }
        public void RemoveAll(Predicate<T> match)
        {
            lock (list)
            {
                list.RemoveAll(match);
                UpdateCache();
            }
        }
        public void RemoveAt(int index)
        {
            lock (list)
            {
                list.RemoveAt(index);
                UpdateCache();
            }
        }
        public int IndexOf(T item)
        {
            return Array.IndexOf(cache, item);
        }
        public bool Contains(T item)
        {
            return IndexOf(item) > -1;
        }
        public void Clear()
        {
            lock (list)
            {
                list.Clear();
                UpdateCache();
            }
        }
        public void CopyTo(T[] target, int count)
        {
            cache.CopyTo(target, count);
        }
        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }
        public IEnumerator<T> GetEnumerator()
        {
            var collection = cache;
            for (int i = 0; i < collection.Length; i++)
                yield return collection[i];
        }
        public void UpdateCache()
        {
            lock (list)
            {
                cache = list.ToArray();
            }
        }

        public SafeList()
        {
            list = new List<T>();
            cache = new T[] { };
        }
    }
}
